scrapy/utils/reqser.py

```diff
@@ -90,10 +92,12 @@ def _find_method(obj, func):
             pass
         else:
             if func_self is obj:
-                name = func.__func__.__name__
-                if _is_private_method(name):
-                    return _mangle_private_name(obj, func, name)
-                return name
+                members = inspect.getmembers(obj, predicate=inspect.ismethod)
```

Comment:
I wonder what's the performance overhead of the change - would it be possible to run a micro-benchmark, e.g. timeit for _find_method, for a typical spider from some project, before and after the change?

Reply:
```python
def test_timeit():
    import timeit

    spider = TestSpider()

    def new_find():
        _find_method(spider, spider._TestSpider__parse_item_private)

    def old_find():
        _old_find_method(spider, spider._TestSpider__parse_item_private)

    new = timeit.timeit(new_find, number=10_000)
    old = timeit.timeit(old_find, number=10_000)
    import pdb; pdb.set_trace()
```

```
(Pdb) new
0.338965222
(Pdb) old
0.010640979000000161
(Pdb) 
```

Reply:
The result is "a float measured in seconds".

Reply:
It does not look like a significant run time to me in absolute terms. Although I wonder if we could optimize for previously-supported scenarios, and only have a significant (in relative terms) increase in run time for the cases which are now supported for the first time.

Reply:
The overhead looks workable, at least for CPython. For PyPy it can be worse, as I recall inspect calls break their jit

Adding cache can complicate things, e.g. when someone decides to add methods dynamically. I guess a fast path can be checking function name first - but it is problematic, because introduces the same issue as fixed by this PR (it is slightly better, but a function named `parse` can be matched to a parse callback even if it is assigned to a different attribute). I'm fine with doing nothing here.

Reply:
@Gallaecio it's possible to use something like this:

```python
def _conditional_find_method(obj, func):
    old_name = _old_find_method(obj, func)
    if _get_method(obj, old_name).__func__ is func.__func__:
        return old_name
    else:
        return  _find_method(obj, func)
```

It will only inspect and traverse object members if the old method detects a false positive (divergency in names). 

Although, I'm not sure it's the best approach because we would be duplicating logic trying to fix a performance problem that we are not sure will exist.

Reply:
Results using the same test case:

```
(Pdb) new
0.32308619000000005
(Pdb) old
0.010666135999999993
(Pdb) cond
0.014966982000000018
(Pdb) 
```

It would be like the old method if the function name matches the object attribute and like the new method if there's a divergence on names.

Reply:
> We would be duplicating logic trying to fix a performance problem that we are not sure will exist.

Indeed. Let’s just use the new implementation, then.